internal/app/service/user_email_service.go

```diff
@@ -0,0 +1,141 @@
+package service
+
+import (
+	"context"
+	"time"
+
+	"github.com/joshsoftware/profile_builder_backend_go/internal/pkg/constants"
+	"github.com/joshsoftware/profile_builder_backend_go/internal/pkg/helpers"
+	"github.com/joshsoftware/profile_builder_backend_go/internal/pkg/specs"
+	"github.com/joshsoftware/profile_builder_backend_go/internal/repository"
+	"go.uber.org/zap"
+)
+
+// UserEmailService is the interface for the user email service
+type UserEmailService interface {
+	SendUserInvitation(ctx context.Context, userID int, request specs.UserSendInvitationRequest) error
+	SendAdminInvitation(ctx context.Context, userID int, request specs.UserSendInvitationRequest) error
+}
+
+// inviteUser sends an email to the user with the invitation link
+func (userService *service) SendUserInvitation(ctx context.Context, userID int, request specs.UserSendInvitationRequest) (err error) {
+	tx, _ := userService.ProfileRepo.BeginTransaction(ctx)
+	defer func() {
+		txErr := userService.ProfileRepo.HandleTransaction(ctx, tx, err)
+		if txErr != nil {
+			err = txErr
+			return
+		}
+	}()
+
+	sendRequest := repository.SendUserInvitationRequest{
+		ProfileID: request.ProfileID,
+	}
+	email, err := userService.UserEmailRepo.GetEmailByProfileID(ctx, sendRequest, tx)
+	if err != nil {
+		zap.S().Error("Error getting email by profile id: ", err)
+		return err
+	}
+
+	now := helpers.GetCurrentISTTime()
+	createSendInvitationRequest := repository.EmailRepo{
+		ProfileID:       request.ProfileID,
+		ProfileComplete: 0,
+		CreatedAt:       now,
+		UpdatedAt:       now,
+		CreatedByID:     userID,
+		UpdatedByID:     userID,
+	}
+
+	err = userService.UserEmailRepo.CreateSendInvitation(ctx, createSendInvitationRequest, tx)
+	if err != nil {
+		zap.S().Error("Error creating send invitation: ", err)
+		return err
+	}
+
+	err = userService.UserLoginRepo.CreateUserAsEmployee(ctx, email, tx)
+	if err != nil {
+		zap.S().Error("Error creating user as employee: ", err)
+		return err
+	}
+
+	go func(email string, profileID int) {
+		for i := 0; i < constants.DefaultMaxRetries; i++ {
+			err := helpers.SendUserInvitation(email, profileID)
+			if err != nil {
```

Comment:
what happen if we get an error 3 times ?
do we have any revert mechanism for invite and User table  for deletion ?


what happen if invite and user table already have entry and we are trying to hit same API?

Reply:
We have to discuss for revert mechanism.
For the invitation it will not allowed to add the record in invitation table as well as the users table  given an error to UI as unable to send email